File: pipeline-scripts/config_parser.py
"""
Configuration parser for AphexPipeline.

This module provides functionality to parse and validate aphex-config.yaml files
against the JSON schema.
"""

import json
import yaml
from dataclasses import dataclass
from typing import List, Optional
from pathlib import Path
from jsonschema import validate, ValidationError


@dataclass
class StackConfig:
    """Configuration for a CDK stack."""
    name: str
    path: str


@dataclass
class TestConfig:
    """Configuration for test execution."""
    commands: List[str]


@dataclass
class EnvironmentConfig:
    """Configuration for a deployment environment."""
    name: str
    region: str
    account: str
    stacks: List[StackConfig]
    tests: Optional[TestConfig] = None


@dataclass
class BuildConfig:
    """Configuration for build stage."""
    commands: List[str]


@dataclass
class AphexConfig:
    """Complete AphexPipeline configuration."""
    version: str
    build: BuildConfig
    environments: List[EnvironmentConfig]


class ConfigParser:
    """Parser for AphexPipeline configuration files."""
    
    def __init__(self, schema_path: str = "aphex-config.schema.json"):
        """
        Initialize the configuration parser.
        
        Args:
            schema_path: Path to the JSON schema file
        """
        self.schema_path = schema_path
        self.schema = self._load_schema()
    
    def _load_schema(self) -> dict:
        """Load the JSON schema from file."""
        schema_file = Path(self.schema_path)
        if not schema_file.exists():
            raise FileNotFoundError(f"Schema file not found: {self.schema_path}")
        
        with open(schema_file, 'r') as f:
            return json.load(f)
    
    def parse(self, config_path: str) -> AphexConfig:
        """
        Parse and validate a configuration file.
        
        Args:
            config_path: Path to the aphex-config.yaml file
            
        Returns:
            Parsed and validated AphexConfig object
            
        Raises:
            FileNotFoundError: If config file doesn't exist
            ValidationError: If config doesn't match schema
            yaml.YAMLError: If YAML is malformed
        """
        config_file = Path(config_path)
        if not config_file.exists():
            raise FileNotFoundError(f"Configuration file not found: {config_path}")
        
        # Load YAML
        with open(config_file, 'r') as f:
            config_data = yaml.safe_load(f)
        
        # Validate against schema
        validate(instance=config_data, schema=self.schema)
        
        # Parse into structured objects
        return self._parse_config(config_data)
    
    def _parse_config(self, config_data: dict) -> AphexConfig:
        """Convert raw config data into AphexConfig object."""
        # Parse build config
        build_config = BuildConfig(
            commands=config_data['build']['commands']
        )
        
        # Parse environments
        environments = []
        for env_data in config_data['environments']:
            # Parse stacks
            stacks = [
                StackConfig(name=stack['name'], path=stack['path'])
                for stack in env_data['stacks']
            ]
            
            # Parse optional tests
            tests = None
            if 'tests' in env_data and 'commands' in env_data['tests']:
                tests = TestConfig(commands=env_data['tests']['commands'])
            
            environments.append(EnvironmentConfig(
                name=env_data['name'],
                region=env_data['region'],
                account=env_data['account'],
                stacks=stacks,
                tests=tests
            ))
        
        return AphexConfig(
            version=config_data['version'],
            build=build_config,
            environments=environments
        )


def parse_config(config_path: str, schema_path: str = "aphex-config.schema.json") -> AphexConfig:
    """
    Convenience function to parse a configuration file.
    
    Args:
        config_path: Path to the aphex-config.yaml file
        schema_path: Path to the JSON schema file
        
    Returns:
        Parsed and validated AphexConfig object
    """
    parser = ConfigParser(schema_path=schema_path)
    return parser.parse(config_path)

File: pipeline-infra/lib/config-parser.ts
import * as fs from 'fs';
import * as yaml from 'js-yaml';
import * as path from 'path';

/**
 * Configuration for a CDK stack.
 */
export interface StackConfig {
  name: string;
  path: string;
}

/**
 * Configuration for test execution.
 */
export interface TestConfig {
  commands: string[];
}

/**
 * Configuration for a deployment environment.
 */
export interface EnvironmentConfig {
  name: string;
  region: string;
  account: string;
  stacks: StackConfig[];
  tests?: TestConfig;
}

/**
 * Configuration for build stage.
 */
export interface BuildConfig {
  commands: string[];
}

/**
 * Complete AphexPipeline configuration.
 */
export interface AphexConfig {
  version: string;
  build: BuildConfig;
  environments: EnvironmentConfig[];
}

/**
 * Parser for AphexPipeline configuration files.
 */
export class ConfigParser {
  /**
   * Parse and validate a configuration file.
   * 
   * @param configPath Path to the aphex-config.yaml file
   * @returns Parsed and validated AphexConfig object
   */
  static parse(configPath: string): AphexConfig {
    // Check if file exists
    if (!fs.existsSync(configPath)) {
      throw new Error(`Configuration file not found: ${configPath}`);
    }

    // Load YAML
    const fileContents = fs.readFileSync(configPath, 'utf8');
    const configData = yaml.load(fileContents) as any;

    // Basic validation
    if (!configData.version) {
      throw new Error('Configuration missing required field: version');
    }
    if (!configData.build || !configData.build.commands) {
      throw new Error('Configuration missing required field: build.commands');
    }
    if (!configData.environments || !Array.isArray(configData.environments)) {
      throw new Error('Configuration missing required field: environments');
    }

    // Parse build config
    const build: BuildConfig = {
      commands: configData.build.commands,
    };

    // Parse environments
    const environments: EnvironmentConfig[] = configData.environments.map((envData: any) => {
      if (!envData.name || !envData.region || !envData.account || !envData.stacks) {
        throw new Error('Environment missing required fields: name, region, account, stacks');
      }

      const stacks: StackConfig[] = envData.stacks.map((stackData: any) => {
        if (!stackData.name || !stackData.path) {
          throw new Error('Stack missing required fields: name, path');
        }
        return {
          name: stackData.name,
          path: stackData.path,
        };
      });

      const tests: TestConfig | undefined = envData.tests?.commands
        ? { commands: envData.tests.commands }
        : undefined;

      return {
        name: envData.name,
        region: envData.region,
        account: envData.account,
        stacks,
        tests,
      };
    });

    return {
      version: configData.version,
      build,
      environments,
    };
  }
}
